src/ESPeasyGPIO.ino

```diff
@@ -0,0 +1,393 @@
+  byte port = Par1 - (unit * 16);
+  uint8_t address = 0x20 + unit;
+  byte IOBankConfigReg = 0xC;
+  if (port > 8)
+  {
+    port = port - 8;
+    IOBankConfigReg++;
+  }
+  // turn this port pullup on
+  Wire.beginTransmission(address);
+  Wire.write(IOBankConfigReg);
+  Wire.endTransmission();
+  Wire.requestFrom(address, (uint8_t)0x1);
+  if (Wire.available())
+  {
+    portvalue = Wire.read();
+    if (Par2 == 1)
+      portvalue |= (1 << (port - 1));
+    else
+      portvalue &= ~(1 << (port - 1));
+
+    // write new IO config
+    Wire.beginTransmission(address);
+    Wire.write(IOBankConfigReg); // IO config register
+    Wire.write(portvalue);
+    Wire.endTransmission();
+  }
+}
+
+//********************************************************************************
+// PCF8574 read
+//********************************************************************************
+//@giig1967g-20181023: changed to int8_t
+int8_t GPIO_PCF_Read(byte Par1)
+{
+  int8_t state = -1;
+  byte unit = (Par1 - 1) / 8;
+  byte port = Par1 - (unit * 8);
+  uint8_t address = 0x20 + unit;
+  if (unit > 7) address += 0x10;
+
+  // get the current pin status
+  Wire.requestFrom(address, (uint8_t)0x1);
+  if (Wire.available())
+  {
+    state = ((Wire.read() & _BV(port - 1)) >> (port - 1));
+  }
+  return state;
+}
+
+uint8_t GPIO_PCF_ReadAllPins(uint8_t address)
+{
+  uint8_t rawState = 0;
+
+  Wire.requestFrom(address, (uint8_t)0x1);
+  if (Wire.available())
+  {
+    rawState =Wire.read();
+  }
+  return rawState;
+}
+
+//********************************************************************************
+// PCF8574 write
+//********************************************************************************
+boolean GPIO_PCF_Write(byte Par1, byte Par2)
+{
+  uint8_t unit = (Par1 - 1) / 8;
+  uint8_t port = Par1 - (unit * 8);
+  uint8_t address = 0x20 + unit;
+  if (unit > 7) address += 0x10;
+
+  //generate bitmask
+  int i = 0;
+  uint8_t portmask = 255;
+  unit = unit * 8 + 1; // calculate first pin
+
+  uint32_t key;
+
+  for(i=0; i<8; i++){
+    key = createKey(PLUGIN_PCF,unit+i);
+
+    if (existPortStatus(key) && globalMapPortStatus[key].mode == PIN_MODE_OUTPUT && globalMapPortStatus[key].state == 0)
+      portmask &= ~(1 << i); //set port i = 0
+  }
+
+  key = createKey(PLUGIN_PCF,Par1);
+
+  if (Par2 == 1)
+    portmask |= (1 << (port-1));
+  else
+    portmask &= ~(1 << (port-1));
+
+  Wire.beginTransmission(address);
+  Wire.write(portmask);
+  Wire.endTransmission();
+
+  return true;
+}
+
+//*********************************************************
+// GPIO_Monitor10xSec:
+// What it does:
+// a) It updates the pinstate structure
+// b) Returns an EVENT in case monitor is enabled
+// c) Does not update the pinstate state if the port is defined in a task, as it will be picked up by thePLUGIN_TEN_PER_SECOND event
+// called from: run10TimesPerSecond in ESPEasy.ino
+//*********************************************************
+
+void GPIO_Monitor10xSec()
+{
```

Comment:
@giig1967g @TD-er 
I understand this routine will be optimized later as we have here place to improve.
Before globalMapPortStatus loop, we should cache (last value, actual value) here all expenders at once to avoid unnecessary communication.
For every PCF it is one register, for MCP 4 registers.
With this idea we can modify these cached registers when needed and use it if change on inputs.
After loop update registers if any changes. 
This will really reduce I2C communication as for every MCP/PCF GPIO bit it is several 8/16/32 times per chip. 
Be aware this routine will try to fetch data every 0,1s. and should be as fast as possible.
What do You think?

Reply:
I totally agree, you should only fetch/write the data for all pins at once.
This can also be done for the internal GPIO's, which is just a matter of reading a 32 bit value (well there are at least 3 values for the ESP internals which you have to combine).
So all actions must be done on a single 32 bit member and also using a bitmask for hiding/selecting what bits are of interest.

But first lets make it all be handled in single functions, so then we can later optimize it. First let it be a working solution.

Reply:
sure. it's true that this function need to be fully optimized.
I have come out with several ideas:
1) as you said, read and write PCF and MCP in blocks of 8 bit
2) limit the monitor function to monitor ONLY the input pins, assuming that the outut pins can change only using a command and therefore the monitor event could be sent inside the command itself.
3) for input pins, use interrupts